src/plugins/util/OverlayPlugin.js

```diff
@@ -2,108 +2,358 @@
+                }
+
+                requestAnimationFrame(() => {
+                    let newX = e.clientX - offsetX;
+                    let newY = e.clientY - offsetY;
+
+                    const modalWidth = modal.offsetWidth;
+                    const modalHeight = modal.offsetHeight;
+
+                    // Constrain x and y to be within the wrapper's bounds
+                    newX = Math.max(wrapperRect.left, Math.min(wrapperRect.right - modalWidth, newX));
+                    newY = Math.max(0, Math.min(wrapperRect.bottom - modalHeight - wrapperRect.top, newY));
+
+                    // Magnetic grid (snap to initial position if close)
+                    if (Math.abs(initialX - newX) < 30 && Math.abs(initialY - newY) < 30) {
+                        newX = initialX;
+                        newY = initialY;
+                    }
+
+                    modal.style.left = `${newX}px`;
+                    modal.style.top = `${newY}px`;
+                });
+            });
+
+            const removeMouseUpListener = this.instance.addListener(document, 'mouseup', () => {
+                modal.style.userSelect = 'initial';
+                isDragging = false;
+                this.position.x = modal.style.left;
+                this.position.y = modal.style.top;
+
+                removeMouseMoveListener();
+                removeMouseUpListener();
+            });
+        });
+    }
+
+    /**
+     * Handles drag start initialization
+     * @param {HTMLElement} modal - Modal element
+     * @returns {Object} Initial x/y coordinates
+     */
+    startDragging(modal) {
+        const {button: toolbarButton} = this.instance.toolbarManager.getButtonById(this.constructor.name);
+        const initialPosition = this.calculateModalPosition(toolbarButton, this.instance.options.toolbarPosition);
+
+        modal.style.position = 'absolute';
+        modal.style.zIndex = ++OverlayPlugin.modalZIndex;
+        modal.style.userSelect = 'none';
+        return initialPosition;
     }
+
+    /*
+     * --------------------------
+     * Positioning & Rendering Section
+     * --------------------------
+     */
+
     /**
-     * Closes overlay and updates toolbar button state
+     * Calculates optimal position for modal display
+     * @param {HTMLElement} triggerElement - Element that triggered the modal
+     * @param {string} toolbarPosition - Toolbar position ('left', 'right')
+     * @returns {Object} Calculated x/y coordinates
      */
-    closeOverlay() {
-        if (this.widget && !this.widget.classList.contains('gm-hidden')) {
-            this.widget.classList.add('gm-hidden');
-            if (this.widget.onclose) {
-                this.widget.onclose();
-            }
+    calculateModalPosition(triggerElement, toolbarPosition) {
+        const triggerRect = triggerElement.getBoundingClientRect();
+        const modalWidth = this.widget.offsetWidth || 200;
+        const modalHeight = this.widget.offsetHeight || 100;
```

Comment:
Maybe these values could be extracted to a CONST at the begining of this file. Similar comment for the +/- 10 values in the next lines of this function